common/types/encoding/codecv0/codecv0.go

```diff
@@ -0,0 +1,461 @@
+	}
+
+	return &daBlock, nil
+}
+
+// Encode serializes the DABlock into a slice of bytes.
+func (b *DABlock) Encode() []byte {
+	bytes := make([]byte, 60)
+	binary.BigEndian.PutUint64(bytes[0:], b.BlockNumber)
+	binary.BigEndian.PutUint64(bytes[8:], b.Timestamp)
+	if b.BaseFee != nil {
+		binary.BigEndian.PutUint64(bytes[40:], b.BaseFee.Uint64())
+	}
+	binary.BigEndian.PutUint64(bytes[48:], b.GasLimit)
+	binary.BigEndian.PutUint16(bytes[56:], b.NumTransactions)
+	binary.BigEndian.PutUint16(bytes[58:], b.NumL1Messages)
+	return bytes
+}
+
+// NewDAChunk creates a new DAChunk from the given encoding.Chunk and the total number of L1 messages popped before.
+func NewDAChunk(chunk *encoding.Chunk, totalL1MessagePoppedBefore uint64) (*DAChunk, error) {
+	var blocks []*DABlock
+	var txs [][]*types.TransactionData
+
+	if chunk == nil {
+		log.Crit("failed to new da chunk, chunk is nil")
+	}
+
+	if len(chunk.Blocks) == 0 {
+		log.Crit("failed to new da chunk, chunk contains no block")
+	}
+
+	for _, block := range chunk.Blocks {
+		b, err := NewDABlock(block, totalL1MessagePoppedBefore)
+		if err != nil {
+			return nil, err
+		}
+		blocks = append(blocks, b)
+		totalL1MessagePoppedBefore += block.NumL1Messages(totalL1MessagePoppedBefore)
+		txs = append(txs, block.Transactions)
+	}
+
+	daChunk := DAChunk{
+		Blocks:       blocks,
+		Transactions: txs,
+	}
+
+	return &daChunk, nil
+}
+
+// Encode serializes the DAChunk into a slice of bytes.
+func (c *DAChunk) Encode() []byte {
+	var chunkBytes []byte
+	chunkBytes = append(chunkBytes, byte(len(c.Blocks)))
+
+	var l2TxDataBytes []byte
+
+	for _, block := range c.Blocks {
+		chunkBytes = append(chunkBytes, block.Encode()...)
+	}
+
+	for _, blockTxs := range c.Transactions {
+		for _, txData := range blockTxs {
+			if txData.Type == types.L1MessageTxType {
+				continue
+			}
+			var txLen [4]byte
+			rlpTxData := encoding.MustConvertTxDataToRLPEncoding(txData)
+			binary.BigEndian.PutUint32(txLen[:], uint32(len(rlpTxData)))
+			l2TxDataBytes = append(l2TxDataBytes, txLen[:]...)
+			l2TxDataBytes = append(l2TxDataBytes, rlpTxData...)
+		}
+	}
+
+	chunkBytes = append(chunkBytes, l2TxDataBytes...)
+	return chunkBytes
+}
+
+// Hash computes the hash of the DAChunk data.
+func (c *DAChunk) Hash() common.Hash {
+	chunkBytes := c.Encode()
+	numBlocks := chunkBytes[0]
+
+	// concatenate block contexts
+	var dataBytes []byte
+	for i := 0; i < int(numBlocks); i++ {
+		// only the first 58 bytes of each BlockContext are needed for the hashing process
+		dataBytes = append(dataBytes, chunkBytes[1+60*i:60*i+59]...)
+	}
+
+	// concatenate l1 and l2 tx hashes
+	for _, blockTxs := range c.Transactions {
+		var l1TxHashes []byte
+		var l2TxHashes []byte
+		for _, txData := range blockTxs {
+			txHash := strings.TrimPrefix(txData.TxHash, "0x")
+			hashBytes, err := hex.DecodeString(txHash)
+			if err != nil {
+				log.Crit("failed to decode tx hash from TransactionData", "hash", txData.TxHash)
```

Comment:
Return error

Reply:
removed in [1edfff9](https://github.com/scroll-tech/scroll/pull/1178/commits/1edfff9e4ac3efb4a33e8e770529370a8f066d64)